### ISSUE TITLE
research/allocator: state-conditioned sequential budget allocation across LLM strategies

### ISSUE THREAD

**Issue by cogco1** [cogco1/MonkeyHub#268]
## Goal

Build the first controlled experiment for **state-conditioned sequential allocation across multiple LLM/Harness strategies**.

The research question is deliberately narrow:

> Given the same exact project/task state, several competing Agent strategies, and a fixed total compute budget, how should the next unit of budget be allocated across strategies as observations accumulate?

This issue does **not** train a foundation model, does not optimize architectural form directly, and does not make token usage part of the evaluator.

Related:
- #260 — native-project / Stage-scoped execution harness
- #32 — ContextPack / DesignTransaction / fresh-session context compilation
- #185 — project-side task state / decisions / Stage handoff
- #240 — presentation-only Budgeted Agentic Loop Allocation framing
- #124 / #236 — prior OCBA / evaluation-budget experiments

---

## Core setup

At one fixed state snapshot (S_t):

```text
                 exact State S_t
                      │
          ┌───────────┼───────────┐
          ↓           ↓           ↓
       Strategy A  Strategy B  Strategy C  ...
          │           │           │
      fresh session fresh session fresh session
          │           │           │
        rollout     rollout      rollout
          │           │           │
          └────── external evaluator ──────┘
                      │
              per-strategy samples
                      │
           mean / variance / n
                      │
             sequential allocator
                      │
        next allocation of rollout budget
```

Important:

> **fork from project state, not continue from conversation memory.**

Every rollout must start from the same exact state snapshot and an independent fresh provider session. Strategies must not see one another's transcripts or results.

---

## What is a "strategy"?

A strategy is a replaceable decision policy for the next action / short horizon.

V0 examples:

- **A — free LLM choice**: model chooses next action from the allowed action set.
- **B — fixed workflow**: deterministic hand-authored sequence/baseline.
- **C — state-ranked actions**: strategy ranks currently legal actions from the compiled state and executes the top choice.
- **D — short lookahead**: strategy considers a small number of candidate actions / one-to-two-step continuations before choosing.

The strategies may use the same foundation model. They differ in policy/context/tool exposure, not model weights.

---

## Phase 1 — keep the statistical object stable

For the first experiment, fix:

```text
foundation model/version
temperature / sampling config
State snapshot
allowed action set
single-rollout token cap
rollout horizon
tool/environment version
evaluator version
```

The allocator should initially allocate **additional independent rollout samples**, not dynamically change the token cap inside one rollout.

Example:

```text
each rollout cap = 4k tokens

initialization:
A × 2
B × 2
C × 2
D × 2

after observations:
allocator decides:
A +1
B +0
C +3
D +2
```

This keeps the per-strategy outcome distribution interpretable.

A later phase may treat token level itself as a fidelity variable:

```text
strategy × token_budget
```

but that is explicitly out of V0.

---

## Sequential allocation starts at round 1

Do not treat OCBA/allocation as a post-processing step after a large fixed pilot.

The loop is:

```text
small balanced initialization
        ↓
observe outcomes
        ↓
update mean / variance / n
        ↓
allocation rule chooses next sample(s)
        ↓
run fresh rollout(s)
        ↓
update statistics
        ↓
repeat until budget/stopping condition
```

The initialization samples are part of the sequential allocation procedure.

---

## Evaluator boundary

The evaluator must be **external to token allocation**.

Do not define:

```text
score = task quality - lambda * token
```

because token is the allocated resource and should remain a budget variable, not part of the outcome definition in V0.

Instead separate:

### Outcome / evaluator

Measures task result independently of budget, e.g.:

- task completed / not completed;
- required preconditions satisfied;
- hard constraints preserved;
- correct target/scope selected;
- no forbidden dependency violation;
- reference-optimal / near-optimal action selected in controlled fixtures;
- final artifact/state passes independent readback;
- optional external quality measure where ground truth exists.

### Budget / cost

Recorded separately:

- tokens;
- wall time;
- tool calls;
- API/compute cost;
- human interruptions.

Primary comparison:

> **At the same total budget, which sequential allocation rule achieves the highest external task outcome / completion rate?**

Secondary comparison:

> **For the same external success criterion, which rule reaches it with less budget?**

---

## First fixture: finite controlled harness

Do not begin with open-ended architectural aesthetics.

Create a small deterministic environment with known state transitions and legal/illegal actions.

Suggested cases:

### Case 1 — provided authoritative source

State:
- exact source already exists;
- required passage not yet read.

Actions:
- ReadProvidedSource
- BM25
- RAG
- Model
- AskHuman

Known structure:
- repeated RAG is unnecessary;
- modeling before reading required evidence has higher failure risk.

### Case 2 — protected dependency

State:
- editable object has a protected relation;
- dependency facts are available but unread.

Actions:
- InspectDependency
- Model
- Repair
- RAG

Known structure:
- direct mutation without dependency inspection is unsafe.

### Case 3 — unresolved design direction

State:
- technical preconditions complete;
- multiple valid directions remain.

Actions:
- LocalStudy
- CoarseModel
- AskHuman
- Inspect

Known structure:
- exploration is legal; hard pruning should not collapse all alternatives.

### Case 4 — local conflict

State:
- candidate exists;
- one bounded conflict is known.

Actions:
- Repair
- ReModel
- RAG
- Inspect

Known structure:
- repair/inspection may dominate unrelated retrieval.

The simulator/environment should expose the same interface to all strategies and retain a hidden reference transition/outcome table only for evaluation.

---

## Harness contract

MonkeyHub side should expose an experiment API approximately like:

```text
state = env.reset(case_id, seed)

proposal = strategy.choose(
    state=state,
    allowed_actions=...
)

result = env.rollout(
    state_snapshot=state,
    proposal=proposal,
    horizon=H,
    token_cap=T,
    fresh_session=true
)

observation = evaluator(result)
```

Every rollout record should retain:

```text
case_id
state_snapshot_id / digest
strategy_id / version
model / provider / version
sampling config
seed
token cap
horizon
chosen action(s)
trajectory
external evaluator result
success / failure reason
actual tokens used          # resource only
wall time                   # resource only
tool calls                  # resource only
```

No strategy receives another strategy's rollout result except through the allocator's aggregate statistics if the experimental condition allows it.

---

## Allocation-rule interface

Algorithm collaborators should be able to work without editing MonkeyHub product internals.

Conceptually:

```text
AllocationState
  strategy_id
  n
  sample_mean
  sample_variance
  remaining_budget
  fixed_sample_cost
  optional parallel capacity

        ↓

AllocationRule

        ↓

NextAllocation
  strategy_id(s)
  additional_rollouts
  stopping_reason / diagnostics
```

V0 baselines:

- equal allocation;
- round robin;
- epsilon-greedy / best-estimate focused;
- classical OCBA-style allocation;
- any collaborator-proposed sequential rule.

No rule is assumed to win.

---

## Primary metrics

At matched total budget:

- external task completion rate;
- probability of correct / near-correct next-action selection where reference truth exists;
- simple regret / opportunity loss in finite fixtures;
- constraint / preserve violation rate;
- erroneous pruning rate;
- rollout count by strategy;
- allocation trajectory;
- stopping behavior.

Report budget separately:

- actual tokens;
- wall time;
- tool calls;
- API/compute usage.

Do not mix token expenditure into the evaluator outcome in V0.

---

## Independence / fairness requirements

- same exact State snapshot for competing strategies;
- fresh provider session per rollout;
- no prior transcript inheritance;
- same model/version and tool/environment build;
- same evaluator;
- same single-rollout token cap and horizon in V0;
- randomized / counterbalanced run order where parallel execution is unavailable;
- retain all failures/timeouts instead of replacing them;
- if true simultaneous parallel execution is available, use isolated sessions/workspaces;
- project state remains read-only during strategy comparison unless the fixture explicitly uses a cloned speculative branch.

---

## Phase 2 — only after V0 works

Explore **strategy × compute fidelity**:

```text
Strategy A @ 2k
Strategy A @ 4k
Strategy A @ 8k
Strategy B @ 2k
...
```

Now token budget changes the policy's behavior distribution and should be modeled explicitly as a fidelity / treatment variable.

Possible directions:
- multi-fidelity ranking-and-selection;
- adaptive test-time compute allocation;
- cost-aware sequential sampling;
- state-conditioned strategy allocation.

Do not claim classical OCBA guarantees automatically extend to this setting.

---

## Research hypothesis

The interesting claim is not:

> OCBA improves architecture.

It is closer to:

> **At a fixed project/task state, adaptive sequential allocation of test-time compute across competing LLM/Harness strategies can use a fixed total budget more effectively than equal allocation or a single fixed strategy.**

A stronger later claim would require real tool execution and architectural tasks.

---

## First implementation slice

1. Reuse current project-side state / fresh-session capability from #32/#185.
2. Implement one immutable experiment `StateSnapshot` representation over existing owners; do not create a second project authority.
3. Implement 3–4 finite controlled cases.
4. Implement A/B/C/D strategy wrappers using the same base model.
5. Prove truly independent fresh-session rollouts from the same snapshot.
6. Implement the external evaluator for the finite cases.
7. Emit standardized rollout JSON/CSV.
8. Implement equal / round-robin baselines.
9. Provide the allocator interface to algorithm collaborators.
10. Integrate the first OCBA-style rule only after the harness produces valid repeated samples.

---

## Acceptance

- [ ] One State snapshot can fork at least four independent fresh LLM strategy sessions.
- [ ] No arm receives another arm's transcript/history.
- [ ] Repeated rollouts produce per-strategy empirical outcome distributions.
- [ ] Evaluator outcome is independent of token usage in V0.
- [ ] Token cap/horizon/model are fixed across arms in V0.
- [ ] Mean/variance/sample count are reproducibly computed from retained outcomes.
- [ ] A sequential allocation rule can request the next rollout(s) without touching project authority.
- [ ] Equal allocation and one OCBA-style allocator run under the same total budget.
- [ ] All failures/timeouts count in the experiment.
- [ ] Allocation trajectories and seeds/config are retained.
- [ ] The experiment can be rerun from the same exact State snapshot after restart.
- [ ] No model training or custom inference model is introduced.

---

## Non-goals

- training/fine-tuning a foundation model;
- building a universal reward model;
- using architectural aesthetics as the first evaluator;
- modifying canonical project state during benchmark arms;
- letting conversation memory serve as experiment state;
- dynamic per-rollout token cap in V0;
- claiming theoretical OCBA guarantees before assumptions are checked;
- claiming novelty before a literature review.

---

## Deliverable for collaborators

Provide a small research package that can be used without reading the MonkeyHub codebase:

```text
README.md
state_schema.json
strategy_interface.py
allocator_interface.py
fixtures/
rollout_results/
benchmark.py
```

The algorithm-side task should reduce to:

> Given current per-strategy sample statistics and remaining budget, return the next allocation.
